# Part 2: Data Cleaning 

The large scraped csv file with 21071 records had many set backs such as a lot of NaN values, douplicates, and formating issues. A few tutorials, which are mentioned later in this section, have been used for parts with Google API and GeoJson sections.

#### Section 1 - Data Cleaning goals: 
    1. Drop Duplicates and NaN values
    2. Format Salary to be displayed as salary per year 
  
#### Section 2 - Make more formating additions/changes to later make maps: 
    1. Generate latitute and longitude coordinates for "Location" column (this was done with Google API) 
    2. Make a GeoJson file to be used for mapping later

In [ ]:
import pandas as pd
import numpy as np
import csv
import re
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
data = pd.read_csv('WebScraper.csv')

In [3]:
data.head()

Unnamed: 0                                              Title Location  \
0           0                     Software Engineer - Internship      NaN   
1           1  Fullstack Software Developer Intern (JavaScrip...      NaN   
2           2                       Civil Engineering Internship      NaN   
3           3                            Backend Engineer Intern      NaN   
4           4                               ENGINEER INTERN (OM)      NaN   

                        Company                   Salary  \
0       United Health - College                      NaN   
1                   GoGoody Inc        $15 - $25 an hour   
2  Thompson Pipe Group - Rialto  $12.25 - $15.00 an hour   
3               Shealth Startup                      NaN   
4               M.C. Dean, Inc.                      NaN   

                                         Describtion  \
0  You’ll need to work in a disciplined approach ...   
1  B.S. or M.S. in computer science or related fi...   
2  CAD software: 2 years (Required).The Thompson ...   
3  Design & develop best of the breed backend ser...   
4  Medical, dental, vision, life and disability i...   

                                        Requirements  Rating  
0                                                NaN     3.7  
1                                JavaScript: 2 years     NaN  
2  Microsoft Office: 5 yearsInterpret engineering...     NaN  
3                                                NaN     NaN  
4                                                NaN     3.7

In [4]:
# drop the "Unnamed:0" column
df = data.loc[:, ~data.columns.str.contains('^Unnamed')]
df.head()

Title Location  \
0                     Software Engineer - Internship      NaN   
1  Fullstack Software Developer Intern (JavaScrip...      NaN   
2                       Civil Engineering Internship      NaN   
3                            Backend Engineer Intern      NaN   
4                               ENGINEER INTERN (OM)      NaN   

                        Company                   Salary  \
0       United Health - College                      NaN   
1                   GoGoody Inc        $15 - $25 an hour   
2  Thompson Pipe Group - Rialto  $12.25 - $15.00 an hour   
3               Shealth Startup                      NaN   
4               M.C. Dean, Inc.                      NaN   

                                         Describtion  \
0  You’ll need to work in a disciplined approach ...   
1  B.S. or M.S. in computer science or related fi...   
2  CAD software: 2 years (Required).The Thompson ...   
3  Design & develop best of the breed backend ser...   
4  Medical, dental, vision, life and disability i...   

                                        Requirements  Rating  
0                                                NaN     3.7  
1                                JavaScript: 2 years     NaN  
2  Microsoft Office: 5 yearsInterpret engineering...     NaN  
3                                                NaN     NaN  
4                                                NaN     3.7

In [5]:
#original data frame count 
print(df.shape)

#count douplicated rows in the data frame 
print(df.duplicated().sum())

#check data type
print(df.dtypes)

(21071, 7)
18890
Title            object
Location         object
Company          object
Salary           object
Describtion      object
Requirements     object
Rating          float64
dtype: object


### Results from df.duplicated() show that there are 18890 douplicates! They must be removed.


In [6]:
result_df=df.drop_duplicates()
result_df.head()

Title Location  \
0                     Software Engineer - Internship      NaN   
1  Fullstack Software Developer Intern (JavaScrip...      NaN   
2                       Civil Engineering Internship      NaN   
3                            Backend Engineer Intern      NaN   
4                               ENGINEER INTERN (OM)      NaN   

                        Company                   Salary  \
0       United Health - College                      NaN   
1                   GoGoody Inc        $15 - $25 an hour   
2  Thompson Pipe Group - Rialto  $12.25 - $15.00 an hour   
3               Shealth Startup                      NaN   
4               M.C. Dean, Inc.                      NaN   

                                         Describtion  \
0  You’ll need to work in a disciplined approach ...   
1  B.S. or M.S. in computer science or related fi...   
2  CAD software: 2 years (Required).The Thompson ...   
3  Design & develop best of the breed backend ser...   
4  Medical, dental, vision, life and disability i...   

                                        Requirements  Rating  
0                                                NaN     3.7  
1                                JavaScript: 2 years     NaN  
2  Microsoft Office: 5 yearsInterpret engineering...     NaN  
3                                                NaN     NaN  
4                                                NaN     3.7

In [7]:
# make sure that 'NaN' is np.nan not a regular string
result_df.loc[:,:].replace('NaN',np.nan,inplace=True)
result_df.head(3)

C:\Users\User\anaconda3\lib\site-packages\pandas\core\frame.py:4172: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


Title Location  \
0                     Software Engineer - Internship      NaN   
1  Fullstack Software Developer Intern (JavaScrip...      NaN   
2                       Civil Engineering Internship      NaN   

                        Company                   Salary  \
0       United Health - College                      NaN   
1                   GoGoody Inc        $15 - $25 an hour   
2  Thompson Pipe Group - Rialto  $12.25 - $15.00 an hour   

                                         Describtion  \
0  You’ll need to work in a disciplined approach ...   
1  B.S. or M.S. in computer science or related fi...   
2  CAD software: 2 years (Required).The Thompson ...   

                                        Requirements  Rating  
0                                                NaN     3.7  
1                                JavaScript: 2 years     NaN  
2  Microsoft Office: 5 yearsInterpret engineering...     NaN

In [8]:
#remove rows that have a NaN in the following columns
newdata=result_df.dropna(axis=0, subset=['Rating','Location','Company','Title','Salary'])
newdata

Title  \
800                         NYC Census 2020 College Aide   
1330   Office of Data Governance and Analysis Intern ...   
1335               College Intern, Administration Office   
1346   Undergraduate Internship Program - Computer Sc...   
1439   Engineer Intern (35 Hour) (Target Class - Air ...   
...                                                  ...   
14365                                    ENGINEER INTERN   
14754                     Information Technology Manager   
15159  Undergraduate Internship/Co-op Program - Cyber...   
15161  Undergraduate Internship/Co-op Program - Digit...   
15514             Intern - Florida Prepaid College Board   

                                            Location  \
800                                    Manhattan, NY   
1330                                  Washington, DC   
1335                                Queens, NY 11101   
1346                                  Washington, DC   
1439              Hartford, CT 06106 (Downtown area)   
...                                              ...   
14365           Cincinnati, OH 45202 (West End area)   
14754  Manhattan, NY 10038 (Financial District area)   
15159                                 Washington, DC   
15161                                 Washington, DC   
15514                                Tallahassee, FL   

                                                 Company  \
800          New York City DEPT OF YOUTH & COMM DEV SRVS   
1330                          Legal Services Corporation   
1335         New York City DEPT OF HEALTH/MENTAL HYGIENE   
1346                         Central Intelligence Agency   
1439   State of Connecticut - Department of Energy & ...   
...                                                  ...   
14365                                 City of Cincinnati   
14754       New York City DEPT OF ENVIRONMENT PROTECTION   
15159                        Central Intelligence Agency   
15161                        Central Intelligence Agency   
15514              Florida State Board of Administration   

                         Salary  \
800     $15.50 - $19.90 an hour   
1330          $14 - $20 an hour   
1335    $15.50 - $19.90 an hour   
1346             $22.42 an hour   
1439             $55,625 a year   
...                         ...   
14365  $58,417 - $78,508 a year   
14754  $85,000 - $95,000 a year   
15159   $21.93 - $26.43 an hour   
15161   $21.93 - $26.43 an hour   
15514               $15 an hour   

                                             Describtion Requirements  Rating  
800    Assist Data team in updating project managemen...          NaN     4.1  
1330   The intern must have experience programming in...          NaN     4.2  
1335   Occupational Safety and Health oversees the ag...          NaN     4.1  
1346   As a Computer Science undergraduate student fo...          NaN     4.3  
1439   General knowledge of basic engineering theorie...          NaN     4.0  
...                                                  ...          ...     ...  
14365  Create and enter database and graphics for com...          NaN     3.9  
14754  NOTE: The following types of experience are no...          NaN     3.8  
15159  As a Cyber Operations Officer Undergraduate In...          NaN     4.3  
15161  As a Digital Targeter Undergraduate Intern for...          NaN     4.3  
15514  The Florida Prepaid College Board is currently...          NaN     4.5  

[81 rows x 7 columns]

In [17]:
newdata=newdata.drop(['Requirements'], axis=1)

# Format the Salary

The original salary column is not **consistent** because: 

    1. There is a range ex: 15 -25
    2. There are special strings & characters such as '$' , 'per month', 'per year'
    3. Some values are per hour while some are per year

I need to convert all the salary to display the pay by year 

#####  I followed this tutorial on Medium: https://www.dataquest.io/blog/data-cleaning-with-python/ which talks about cleaning range data

In [10]:
#this shows the different formats this column has
newdata['Salary'].value_counts()

$20 an hour                   8
$21.93 - $26.43 an hour       8
$15.50 - $19.90 an hour       6
$21 - $24 an hour             5
$15 an hour                   5
$13 - $17 an hour             3
$12 an hour                   3
$800 a month                  2
$300 a week                   2
$61,216 a year                2
$60,000 - $70,000 a year      2
$18 an hour                   2
$13 - $23 an hour             2
$55,625 a year                2
$14 - $15 an hour             2
$17 - $19 an hour             2
$14 - $20 an hour             2
$22.42 an hour                2
$15 - $22 an hour             1
$11 an hour                   1
$17 - $22 an hour             1
$21.21 - $22.43 an hour       1
$102,663 - $170,800 a year    1
$17.77 an hour                1
$16.02 - $20.72 an hour       1
$40,000 - $60,000 a year      1
$58,417 - $78,508 a year      1
$16 - $22 an hour             1
$35,119 - $93,638 a year      1
$12 - $14 an hour             1
$48,000 - $55,000 a year      1
$85,000 

In [19]:
# remove $ and 'per year' and commas
newdata.loc[:,'Salary'] = newdata.loc[:,'Salary'].str.replace(r'$', '')
#newdata.loc[:,'Salary'] = newdata.loc[:,'Salary'].str.replace(r'a year', '')
newdata.loc[:,'Salary'] = newdata.loc[:,'Salary'].str.replace(r',', '')

In [20]:
def clean_split_dates(row):
    
    # find initial salary column
    initial_salary= row['Salary']
    
    # Split original salary into two elements if "-" is found
    split_salary = initial_salary.split('-')
    
    #split2 =initial_date.split('an hour')
    
     # If a "-"  is found, split_salary will contain a list with at least two items
    if len(split_salary) > 1:
        final_salary= split_salary[0] #this only keeps the 1st item ex: 12-14 -- only keeps 12
        
    # If no "-" is found, split_date will just contain the original 1 salary item
    else:
        final_salary = initial_salary
    return final_salary

# Assign the results of "clean_split_dates" to the 'Date' column.
# We want Pandas to go row-wise so we set "axis=1". We would use "axis=0" for column-wise.
newdata.loc[:,'Salary'] = newdata.apply(lambda row: clean_split_dates(row), axis=1)
newdata.loc[:,'Salary'].value_counts()

20 an hour       8
21.93            8
15.50            6
15 an hour       5
13               5
21               5
14               4
12 an hour       3
17               3
18 an hour       2
61216 a year     2
55625 a year     2
22.42 an hour    2
300 a week       2
60000            2
800 a month      2
40000            1
51535            1
44279 a year     1
600 a week       1
12               1
102663           1
26.28            1
15               1
21.21            1
65000            1
58417            1
17 an hour       1
88651            1
17.77 an hour    1
35119            1
85000            1
16.02            1
11 an hour       1
16               1
48000            1
Name: Salary, dtype: int64

In [21]:
anhour =newdata[newdata.Salary.str.contains('an hour',case=False)]
anhour

Title  \
1346   Undergraduate Internship Program - Computer Sc...   
3544     Internship - Project Engineer Summer Intern- HH   
3590                                              Intern   
4194         Software Engineering Intern - Melbourne, FL   
4295                                              Intern   
4336                       Summer Engineering Internship   
4497     Internship - Project Engineer Summer Intern- BD   
4794                         Healthcare Analytics Intern   
5117         Software Engineering Intern - Melbourne, FL   
5808                     Topological Data Analyst Intern   
6093                   ITS DevOps Engineer Intern (4920)   
6152                   ITS DevOps Engineer Intern (4920)   
6534                   ITS Web Development Intern (4920)   
6543                   ITS Web Development Intern (4920)   
9446           2020 SUMMER IT STUDENT INTERN - TEMPORARY   
9928                                           IT Intern   
10770                                          IT Intern   
11003                       Administrative Intern (BEST)   
11599               Information Technology Summer Intern   
12076  Undergraduate Internship/Co-op Program - Compu...   
14071                       FALL INTERN - SUSTAINABILITY   
14216                                          IT Intern   
15514             Intern - Florida Prepaid College Board   

                                              Location  \
1346                                    Washington, DC   
3544                                       Chicago, IL   
3590                            College Park, MD 20740   
4194                               Melbourne, FL 32934   
4295                            College Park, MD 20740   
4336                                   Vista, CA 92081   
4497                                       Chicago, IL   
4794                              Brookfield, WI 53005   
5117                               Melbourne, FL 32934   
5808                                        Dayton, OH   
6093   Oakland, CA 94607 (Acorn-Acorn Industrial area)   
6152   Oakland, CA 94607 (Acorn-Acorn Industrial area)   
6534   Oakland, CA 94607 (Acorn-Acorn Industrial area)   
6543   Oakland, CA 94607 (Acorn-Acorn Industrial area)   
9446                                          Cary, NC   
9928                 Chicago, IL 60603 (The Loop area)   
10770                       Farmington Hills, MI 48331   
11003                                       Downey, CA   
11599                               Columbia, SC 29223   
12076                                   Washington, DC   
14071                                      Phoenix, AZ   
14216                       Farmington Hills, MI 48331   
15514                                  Tallahassee, FL   

                                                Company         Salary  \
1346                        Central Intelligence Agency  22.42 an hour   
3544                     IHC Construction Companies LLC     18 an hour   
3590                                     Fraunhofer USA     12 an hour   
4194                                Avidyne Corporation     20 an hour   
4295                                     Fraunhofer USA     12 an hour   
4336                                    Flux Power, Inc     17 an hour   
4497                     IHC Construction Companies LLC     18 an hour   
4794                                           Milliman     12 an hour   
5117                                Avidyne Corporation     20 an hour   
5808                                                KBR     20 an hour   
6093   University of California Office of the President     20 an hour   
6152   University of California Office of the President     20 an hour   
6534   University of California Office of the President     20 an hour   
6543   University of California Office of the President     20 an hour   
9446                                   Town of Cary, NC     15 an hour   
9928                             

In [22]:
anhour.loc[:,'Salary']= anhour.loc[:,'Salary'].str.replace('an hour', '') #sometimes put .str to test at first
anhour.loc[:,'Salary'] = anhour.loc[:,'Salary'].astype(float)
anhour.head()

C:\Users\User\anaconda3\lib\site-packages\pandas\core\indexing.py:965: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


Title  \
1346  Undergraduate Internship Program - Computer Sc...   
3544    Internship - Project Engineer Summer Intern- HH   
3590                                             Intern   
4194        Software Engineering Intern - Melbourne, FL   
4295                                             Intern   

                    Location                         Company  Salary  \
1346          Washington, DC     Central Intelligence Agency   22.42   
3544             Chicago, IL  IHC Construction Companies LLC   18.00   
3590  College Park, MD 20740                  Fraunhofer USA   12.00   
4194     Melbourne, FL 32934             Avidyne Corporation   20.00   
4295  College Park, MD 20740                  Fraunhofer USA   12.00   

                                            Describtion  Rating  
1346  As a Computer Science undergraduate student fo...     4.3  
3544  Applicant should have good organizational skil...     4.3  
3590  Opportunities to develop scientific solutions ...     3.0  
4194  The company has set a new standard in pilot-fr...     4.2  
4295  Chances to apply industry best practices in so...     3.0

In [23]:
anhour.loc[:,'Salary'] *= 2080
anhour.head()

C:\Users\User\anaconda3\lib\site-packages\pandas\core\indexing.py:965: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


Title  \
1346  Undergraduate Internship Program - Computer Sc...   
3544    Internship - Project Engineer Summer Intern- HH   
3590                                             Intern   
4194        Software Engineering Intern - Melbourne, FL   
4295                                             Intern   

                    Location                         Company   Salary  \
1346          Washington, DC     Central Intelligence Agency  46633.6   
3544             Chicago, IL  IHC Construction Companies LLC  37440.0   
3590  College Park, MD 20740                  Fraunhofer USA  24960.0   
4194     Melbourne, FL 32934             Avidyne Corporation  41600.0   
4295  College Park, MD 20740                  Fraunhofer USA  24960.0   

                                            Describtion  Rating  
1346  As a Computer Science undergraduate student fo...     4.3  
3544  Applicant should have good organizational skil...     4.3  
3590  Opportunities to develop scientific solutions ...     3.0  
4194  The company has set a new standard in pilot-fr...     4.2  
4295  Chances to apply industry best practices in so...     3.0

In [24]:
#convert weekly salary to a yearly one
aweek =newdata[newdata.loc[:,'Salary'].str.contains('a week',case=False)]
aweek.loc[:,'Salary']= aweek.loc[:,'Salary'].str.replace('a week', '')
aweek.loc[:,'Salary'] = aweek.loc[:,'Salary'].astype(float)
aweek.loc[:,'Salary'] *= 52
aweek

C:\Users\User\anaconda3\lib\site-packages\pandas\core\indexing.py:965: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


Title                  Location  \
4790  SUMMER INTERNSHIP - Computer Science  Fredericksburg, VA 22408   
5180    Acid Mine Drainage Software Intern            Pittsburgh, PA   
6500    Acid Mine Drainage Software Intern            Pittsburgh, PA   

                  Company   Salary  \
4790         IST Research  31200.0   
5180  Conservation Legacy  15600.0   
6500  Conservation Legacy  15600.0   

                                            Describtion  Rating  
4790  The ability to self-teach complex material by ...     4.8  
5180  Term of positions is 15 weeks; full-time 40 hr...     3.5  
6500  Assist in the completion of the AMDTreat softw...     3.5

In [25]:
ayear =newdata[newdata.loc[:,'Salary'].str.contains('a year',case=False)]
ayear.loc[:,'Salary']= ayear.loc[:,'Salary'].str.replace('a year', '')
ayear.loc[:,'Salary'] = ayear.loc[:,'Salary'].astype(float)
ayear.head()

C:\Users\User\anaconda3\lib\site-packages\pandas\core\indexing.py:965: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


Title  \
1439  Engineer Intern (35 Hour) (Target Class - Air ...   
4293  Engineer Intern (35 Hour) (Land & Water Resour...   
7417  Undergraduate Internship/Co-op Program - Multi...   
7751                 Enterprise Infrastructure Engineer   
7786                 Enterprise Infrastructure Engineer   

                                Location  \
1439  Hartford, CT 06106 (Downtown area)   
4293  Hartford, CT 06106 (Downtown area)   
7417                      Washington, DC   
7751                      Middlebury, VT   
7786                      Middlebury, VT   

                                                Company   Salary  \
1439  State of Connecticut - Department of Energy & ...  55625.0   
4293  State of Connecticut - Department of Energy & ...  55625.0   
7417                        Central Intelligence Agency  44279.0   
7751                                 Middlebury College  61216.0   
7786                                 Middlebury College  61216.0   

                                            Describtion  Rating  
1439  General knowledge of basic engineering theorie...     4.0  
4293  General knowledge of basic engineering theorie...     4.0  
7417  As an Undergraduate Intern, you are expected t...     4.3  
7751  Furthermore, the engineer proactively monitors...     4.0  
7786  Student workers, interns, and representatives ...     4.0

In [27]:
updated=pd.concat([ayear,aweek,anhour],ignore_index=True)
updated.head()

Title  \
0  Engineer Intern (35 Hour) (Target Class - Air ...   
1  Engineer Intern (35 Hour) (Land & Water Resour...   
2  Undergraduate Internship/Co-op Program - Multi...   
3                 Enterprise Infrastructure Engineer   
4                 Enterprise Infrastructure Engineer   

                             Location  \
0  Hartford, CT 06106 (Downtown area)   
1  Hartford, CT 06106 (Downtown area)   
2                      Washington, DC   
3                      Middlebury, VT   
4                      Middlebury, VT   

                                             Company   Salary  \
0  State of Connecticut - Department of Energy & ...  55625.0   
1  State of Connecticut - Department of Energy & ...  55625.0   
2                        Central Intelligence Agency  44279.0   
3                                 Middlebury College  61216.0   
4                                 Middlebury College  61216.0   

                                         Describtion  Rating  
0  General knowledge of basic engineering theorie...     4.0  
1  General knowledge of basic engineering theorie...     4.0  
2  As an Undergraduate Intern, you are expected t...     4.3  
3  Furthermore, the engineer proactively monitors...     4.0  
4  Student workers, interns, and representatives ...     4.0

In [29]:
updated.to_csv('CleanerData1.csv', encoding='utf-8')

# Getting Latitude and Longitude Coordinates 

The goal of getting latitute and longitude was to be able to map out the locations of the companies from my data file. There were a few ways to do it online that I saw but the one that made most sense had to do with using Google APIs which automatically genarate lat. and long. coordinates from the dataframe. 

This is the tutorial I followed: https://towardsdatascience.com/how-to-generate-lat-and-long-coordinates-from-an-address-column-using-pandas-and-googlemaps-api-d66b2720248d


In [35]:
from googlemaps import Client as GoogleMaps
import pandas as pd 

In [36]:
gmaps = GoogleMaps('AIzaSyBjKI8xHvE1pcqAlE1Mfxfzdu18SsX20hQ')

In [40]:
coordinates= pd.read_csv("CleanerData3.csv")
coordinates= coordinates.loc[:, ~coordinates.columns.str.contains('^Unnamed')]
coordinates.head()

Title  \
0  Engineer Intern (35 Hour) (Target Class - Air ...   
1  Engineer Intern (35 Hour) (Land & Water Resour...   
2  Undergraduate Internship/Co-op Program - Multi...   
3                 Enterprise Infrastructure Engineer   
4                 Enterprise Infrastructure Engineer   

                             Location  \
0  Hartford, CT 06106 (Downtown area)   
1  Hartford, CT 06106 (Downtown area)   
2                      Washington, DC   
3                      Middlebury, VT   
4                      Middlebury, VT   

                                             Company   Salary  \
0  State of Connecticut - Department of Energy & ...  55625.0   
1  State of Connecticut - Department of Energy & ...  55625.0   
2                        Central Intelligence Agency  44279.0   
3                                 Middlebury College  61216.0   
4                                 Middlebury College  61216.0   

                                         Describtion  Rating  
0  General knowledge of basic engineering theorie...     4.0  
1  General knowledge of basic engineering theorie...     4.0  
2  As an Undergraduate Intern, you are expected t...     4.3  
3  Furthermore, the engineer proactively monitors...     4.0  
4  Student workers, interns, and representatives ...     4.0

In [41]:
#create 2 empty columns
coordinates['long'] = ""
coordinates['lat'] = ""

In [42]:
for x in range(len(coordinates)):
    geocode_result = gmaps.geocode(coordinates['Location'][x])
    coordinates['lat'][x] = geocode_result[0]['geometry']['location']['lat']
    coordinates['long'][x] = geocode_result[0]['geometry']['location']['lng']
coordinates.head()

C:\Users\User\anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\User\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


Title  \
0  Engineer Intern (35 Hour) (Target Class - Air ...   
1  Engineer Intern (35 Hour) (Land & Water Resour...   
2  Undergraduate Internship/Co-op Program - Multi...   
3                 Enterprise Infrastructure Engineer   
4                 Enterprise Infrastructure Engineer   

                             Location  \
0  Hartford, CT 06106 (Downtown area)   
1  Hartford, CT 06106 (Downtown area)   
2                      Washington, DC   
3                      Middlebury, VT   
4                      Middlebury, VT   

                                             Company   Salary  \
0  State of Connecticut - Department of Energy & ...  55625.0   
1  State of Connecticut - Department of Energy & ...  55625.0   
2                        Central Intelligence Agency  44279.0   
3                                 Middlebury College  61216.0   
4                                 Middlebury College  61216.0   

                                         Describtion  Rating     long      lat  
0  General knowledge of basic engineering theorie...     4.0 -72.6772  41.7656  
1  General knowledge of basic engineering theorie...     4.0 -72.6772  41.7656  
2  As an Undergraduate Intern, you are expected t...     4.3 -77.0369  38.9072  
3  Furthermore, the engineer proactively monitors...     4.0 -73.1673  44.0153  
4  Student workers, interns, and representatives ...     4.0 -73.1673  44.0153

In [47]:
#original data frame count 
print(coordinates.shape)

#count douplicated rows in the data frame 
print(coordinates.duplicated().sum())

#check data type
print(coordinates.dtypes)

#coordinates.to_csv('CleanedFinal.csv')

(31, 8)
0
Title           object
Location        object
Company         object
Salary         float64
Describtion     object
Rating         float64
long            object
lat             object
dtype: object


In [48]:
# convert lat-long to floats and change address from ALL CAPS to Regular Capitalization
coordinates['lat'] = coordinates['lat'].astype(float)
coordinates['long'] = coordinates['long'].astype(float)
coordinates['Location'] = coordinates['Location'].str.title()

In [49]:
print(coordinates.dtypes)

Title           object
Location        object
Company         object
Salary         float64
Describtion     object
Rating         float64
long           float64
lat            float64
dtype: object


In [50]:
coordinates.to_csv('CleanedFinal.csv')

# Converting Location information into GEOJSON to make Maps 

Since I didnt have experience with GeoJson, but still wanted to make a map which required this JSON format, I found a tutorial which goes over how to make GeoJson and provides a few reusable functions 

Tutorial: https://github.com/gboeing/urban-data-science/blob/master/17-Leaflet-Web-Mapping/leaflet-simple-demo/pandas-to-geojson.ipynb

In [132]:
coordinates.head()

Unnamed: 0                                              Title  \
0           0  Engineer Intern (35 Hour) (Target Class - Air ...   
1           1  Engineer Intern (35 Hour) (Land & Water Resour...   
2           2  Undergraduate Internship/Co-op Program - Multi...   
3           3                 Enterprise Infrastructure Engineer   
4           4                 Enterprise Infrastructure Engineer   

                             Location  \
0  Hartford, CT 06106 (Downtown area)   
1  Hartford, CT 06106 (Downtown area)   
2                      Washington, DC   
3                      Middlebury, VT   
4                      Middlebury, VT   

                                             Company   Salary  \
0  State of Connecticut - Department of Energy & ...  55625.0   
1  State of Connecticut - Department of Energy & ...  55625.0   
2                        Central Intelligence Agency  44279.0   
3                                 Middlebury College  61216.0   
4                                 Middlebury College  61216.0   

                                         Describtion  Requirements  Rating  \
0  General knowledge of basic engineering theorie...           NaN     4.0   
1  General knowledge of basic engineering theorie...           NaN     4.0   
2  As an Undergraduate Intern, you are expected t...           NaN     4.3   
3  Furthermore, the engineer proactively monitors...           NaN     4.0   
4  Student workers, interns, and representatives ...           NaN     4.0   

      long      lat  
0 -72.6772  41.7656  
1 -72.6772  41.7656  
2 -77.0369  38.9072  
3 -73.1673  44.0153  
4 -73.1673  44.0153

In [161]:
#original data frame count 
print(coordinates.shape)

#count douplicated rows in the data frame 
print(coordinates.duplicated().sum())

#check data type
print(coordinates.dtypes)

(31, 10)
0
Unnamed: 0        int64
Title            object
Location         object
Company          object
Salary          float64
Describtion      object
Requirements    float64
Rating          float64
long             object
lat              object
dtype: object


In [164]:
# convert lat-long to floats and change address from ALL CAPS to Regular Capitalization
coordinates['lat'] = coordinates['lat'].astype(float)
coordinates['long'] = coordinates['long'].astype(float)
coordinates['Location'] = coordinates['Location'].str.title()

In [231]:
coordinates

Unnamed: 0                                              Title  \
0            0  Engineer Intern (35 Hour) (Target Class - Air ...   
1            1  Engineer Intern (35 Hour) (Land & Water Resour...   
2            2  Undergraduate Internship/Co-op Program - Multi...   
3            3                 Enterprise Infrastructure Engineer   
4            4                 Enterprise Infrastructure Engineer   
5            5               SUMMER INTERNSHIP - Computer Science   
6            6                 Acid Mine Drainage Software Intern   
7            7                 Acid Mine Drainage Software Intern   
8            8  Undergraduate Internship Program - Computer Sc...   
9            9    Internship - Project Engineer Summer Intern- HH   
10          10                                             Intern   
11          11        Software Engineering Intern - Melbourne, FL   
12          12                                             Intern   
13          13                      Summer Engineering Internship   
14          14    Internship - Project Engineer Summer Intern- BD   
15          15                        Healthcare Analytics Intern   
16          16        Software Engineering Intern - Melbourne, FL   
17          17                    Topological Data Analyst Intern   
18          18                  ITS DevOps Engineer Intern (4920)   
19          19                  ITS DevOps Engineer Intern (4920)   
20          20                  ITS Web Development Intern (4920)   
21          21                  ITS Web Development Intern (4920)   
22          22          2020 SUMMER IT STUDENT INTERN - TEMPORARY   
23          23                                          IT Intern   
24          24                                          IT Intern   
25          25                       Administrative Intern (BEST)   
26          26               Information Technology Summer Intern   
27          27  Undergraduate Internship/Co-op Program - Compu...   
28          28                       FALL INTERN - SUSTAINABILITY   
29          29                                          IT Intern   
30          30             Intern - Florida Prepaid College Board   

                                           Location  \
0                Hartford, Ct 06106 (Downtown Area)   
1                Hartford, Ct 06106 (Downtown Area)   
2                                    Washington, Dc   
3                                    Middlebury, Vt   
4                                    Middlebury, Vt   
5                          Fredericksburg, Va 22408   
6                                    Pittsburgh, Pa   
7                                    Pittsburgh, Pa   
8                                    Washington, Dc   
9                                       Chicago, Il   
10                           College Park, Md 20740   
11                              Melbourne, Fl 32934   
12                           College Park, Md 20740   
13                                  Vista, Ca 92081   
14                                      Chicago, Il   
15                             Brookfield, Wi 53005   
16                              Melbourne, Fl 32934   
17                                       Dayton, Oh   
18  Oakland, Ca 94607 (Acorn-Acorn Industrial Area)   
19  Oakland, Ca 94607 (Acorn-Acorn Industrial Area)   
20  Oakland, Ca 94607 (Acorn-Acorn Industrial Area)   
21  Oakland, Ca 94607 (Acorn-Acorn Industrial Area)   
22                                         Cary, Nc   
23                Chicago, Il 60603 (The Loop Area)   
24                       Farmington Hills, Mi 48331   
25                                       Downey, Ca   
26                               Columbia, Sc 29223   
27                                   Washington, Dc   
28                                      Phoenix, Az   
29                       Farmington Hills, Mi 48331   
30                                  Tallahassee, Fl   

                                  

In [194]:
# we don't need all those columns - only keep useful ones
useful_cols = ['lat', 'long','Salary','Rating','Location']
df_subset = coordinates[useful_cols]

In [195]:
# drop any rows that lack lat/long data
df_geo = df_subset.dropna(subset=['lat', 'long'], axis=0, inplace=False)

print('We have {} geotagged rows'.format(len(df_geo)))
df_geo.tail()

We have 31 geotagged rows


lat        long   Salary  Rating                    Location
26  34.105838  -80.920774  22880.0     2.9          Columbia, Sc 29223
27  38.907192  -77.036871  46633.6     4.3              Washington, Dc
28  33.448377 -112.074037  31200.0     3.5                 Phoenix, Az
29  42.508645  -83.410983  31200.0     3.1  Farmington Hills, Mi 48331
30  30.438256  -84.280733  31200.0     4.5             Tallahassee, Fl

In [197]:
# what is the distribution of issue types?
df_geo['Location'].value_counts()

Oakland, Ca 94607 (Acorn-Acorn Industrial Area)    4
Washington, Dc                                     3
Melbourne, Fl 32934                                2
Chicago, Il                                        2
Pittsburgh, Pa                                     2
Farmington Hills, Mi 48331                         2
Hartford, Ct 06106 (Downtown Area)                 2
Middlebury, Vt                                     2
College Park, Md 20740                             2
Downey, Ca                                         1
Brookfield, Wi 53005                               1
Vista, Ca 92081                                    1
Chicago, Il 60603 (The Loop Area)                  1
Fredericksburg, Va 22408                           1
Cary, Nc                                           1
Columbia, Sc 29223                                 1
Phoenix, Az                                        1
Tallahassee, Fl                                    1
Dayton, Oh                                    

In [232]:
coordinates.to_csv('CleanedFinal1.csv')

### The reusable function below was taken from the same tutorial to make the final GEOJson file

In [ ]:
import pandas as json

def df_to_geojson(df, properties, lat='lat', lon='long'):
    """
    Turn a dataframe containing point data into a geojson formatted python dictionary
    
    df : the dataframe to convert to geojson
    properties : a list of columns in the dataframe to turn into geojson feature properties
    lat : the name of the column in the dataframe that contains latitude data
    lon : the name of the column in the dataframe that contains longitude data
    """
    
    # create a new python dict to contain our geojson data, using geojson format
    geojson = {'type':'FeatureCollection', 'features':[]}

    # loop through each row in the dataframe and convert each row to geojson format
    for _, row in df.iterrows():
        # create a feature template to fill in
        feature = {'type':'Feature',
                   'properties':{},
                   'geometry':{'type':'Point',
                               'coordinates':[]}}

        # fill in the coordinates
        feature['geometry']['coordinates'] = [row[lon],row[lat]]

        # for each column, get the value and add it as a new feature property
        for prop in properties:
            feature['properties'][prop] = row[prop]
        
        # add this feature (aka, converted dataframe row) to the list of features inside our dict
        geojson['features'].append(feature)
    
    return geojson

In [ ]:
import geojson
import json

#which useful properties to save to geojson file
useful_columns = ['Rating', 'Salary','Location']
geojson_dict = df_to_geojson(df_geo, properties=useful_columns)
geojson_str = json.dumps(geojson_dict, indent=2)

In [ ]:
# save the geojson result to a file
output_filename = 'dataset2.json'
with open(output_filename, 'w') as output_file:
    output_file.write('var dataset = {};'.format(geojson_str))
    